drivers/interrupt_controller/intc_plic.c

```diff
@@ -208,43 +241,67 @@ static void plic_irq_handler(const void *arg)
 /**
  * @brief Initialize the Platform Level Interrupt Controller
  *
+ * @param dev PLIC device struct
+ *
  * @retval 0 on success.
  */
-static int plic_init(void)
+static int plic_init(const struct device *dev)
 {
-
-	volatile uint32_t *en = (volatile uint32_t *)PLIC_IRQ_EN;
-	volatile uint32_t *prio = (volatile uint32_t *)PLIC_PRIO;
-	volatile struct plic_regs_t *regs =
-	    (volatile struct plic_regs_t *)PLIC_REG;
-	int i;
+	const struct plic_config *config = dev->config;
+	volatile uint32_t *en = (volatile uint32_t *) config->irq_en;
+	volatile uint32_t *prio = (volatile uint32_t *) config->prio;
+	volatile struct plic_regs_t *regs = (volatile struct plic_regs_t *) config->reg;
 
 	/* Ensure that all interrupts are disabled initially */
-	for (i = 0; i < PLIC_EN_SIZE; i++) {
+	for (uint32_t i = 0; i < get_plic_enabled_size(dev); i++) {
 		*en = 0U;
 		en++;
 	}
 
 	/* Set priority of each interrupt line to 0 initially */
-	for (i = 0; i < PLIC_IRQS; i++) {
+	for (uint32_t i = 0; i < config->num_irqs; i++) {
 		*prio = 0U;
 		prio++;
 	}
 
 	/* Set threshold priority to 0 */
 	regs->threshold_prio = 0U;
 
-	/* Setup IRQ handler for PLIC driver */
-	IRQ_CONNECT(RISCV_MACHINE_EXT_IRQ,
-		    0,
-		    plic_irq_handler,
-		    NULL,
-		    0);
-
-	/* Enable IRQ for PLIC driver */
-	irq_enable(RISCV_MACHINE_EXT_IRQ);
+	/* Configure IRQ for PLIC driver */
+	config->irq_config_func();
 
 	return 0;
 }
 
-SYS_INIT(plic_init, PRE_KERNEL_1, CONFIG_INTC_INIT_PRIORITY);
+#define PLIC_INTC_IRQ_FUNC_DECLARE(n) static void plic_irq_config_func_##n(void)
+
+#define PLIC_INTC_IRQ_FUNC_DEFINE(n)                                                               \
+	static void plic_irq_config_func_##n(void)                                                 \
+	{                                                                                          \
+		IRQ_CONNECT(DT_INST_IRQN(n), 0, plic_irq_handler, DEVICE_DT_INST_GET(n), 0);       \
+		irq_enable(DT_INST_IRQN(n));                                                       \
+	}
+
+#define PLIC_INTC_INIT_TRIG_0(n) 0
+#define PLIC_INTC_INIT_TRIG_1(n) DT_INST_REG_ADDR_BY_NAME(n, trig)
+#define PLIC_INTC_INIT_TRIG(n) \
```

Comment:
Assuming that trigger type registers are mandatory, can't we simply derive the exact instance-specific address of `trig` from the value of `num_irqs`? Probably don't even need a DT binding or Kconfig value in that case.